week3/threesum.cc

```diff
@@ -0,0 +1,160 @@
+/* threesum.cc
+ * author: connie
+ *
+ * THREE SUM
+ * Given an input array S of integers, find all unique triplets (i.e. array of 3
+ * values from S) for which the values of the triplets add to 0.
+ * Ex. S = {0,1,1,-2,-1} returns {{0,1,-1},{1,1,-2}}
+ *
+ */
+#include "twosum.h"
+#include <iostream>
+#include <sstream>
+#include <cassert>
+#include <vector>
+#include <map>
+
+using namespace std;
+
+//== DECLARATIONS
+TwoSum twosum;
+vector<int> build_arr(vector<int>& v, int i); // helper
+vector<vector<int>> getThreeSum(vector<int>& v);
+void print_answer(vector<vector<int>> v); // for debugging
+void print_arr(vector<int> v); // for debugging
+
+//== DEFINITIONS
+//#define DEBUG
+
+//== FUNCTIONS
+/*
+ * for debugging
+ */
+void print_arr(vector<int> v) {
+  if (v.size() == 0) { cout << "empty array!" << endl; return; }
+
+  for (int i = 0; i < v.size(); i++) {
+    cout << v[i] << " ";
+  }
+  cout << " " << endl;
+}
+
+void print_arr_of_arr(vector<vector<int>> v) {
+  if (v.size() == 0) { cout << "empty array!" << endl; return; }
+
+  for (int i = 0; i < v.size(); i++) {
+    cout << "triplet " << i << ": ";
+    for (int j = 0; j < 3; j++) {
+      cout << v[i][j] << " ";
+    }
+    cout << " " << endl;
+  }
+  cout << " " << endl;
+}
+
+/*
+ * vector<int> build_arr(vector<int>& v, int i)
+ * inputs: v - array to build a new array from
+           i - index to exclude from the new array built from v
+ * output: a new array that has every value in v except the one at v[i]
+ * notes: could have put this in the main function but wanted to declutter
+ */
+vector<int> build_arr(vector<int>& v, int i) {
+  vector<int> new_v;
+  new_v.assign(v.begin(),v.end());
+  new_v.erase(new_v.begin()+i);
+
+  return new_v;
+}
+
+/*
+ * vector<vector<int>> getThreeSum(vector<int>& v)
+ * input: v - the array to find triplet threesum arrays from
+ * output: an array of triplets (i.e. arrays w/ 3 values) with the desired
+           threesum from the input array
+ * notes: the main solution function
+ */
+vector<vector<int>> getThreeSum(vector<int>& v)
+{
+  // if v doesn't fulfill basic threesum requirements, return an empty arr
+  if (v.size() < 3) return {};
+
+  // create a map to hold existing threesums
+  map<int,int> threesums;
+  vector<vector<int>> threesum_result = {};
+  // for twosum operations
+  vector<int> twosum_arr; int twosum_target;
+  vector<int> twosum_result;
+
+  // next, iterate through the array looking for threesums
+  for (int i = 0; i < v.size(); i++) {
+
+    // first make sure it is not a duplicate using our map
+    if (threesums.find(v[i]) == threesums.end()) {
+      twosum_target = v[i]; // NOTE: we want the additive inverse of this
+      twosum_arr = build_arr(v,i); // build twosum_arr w/ every value except v[i]
```

Comment:
Just a tiny optimization here. Maybe here you don't need to copy every value, just the ones after "i" because you have already checked the previous ones and if there were a triplet including both the current and a previous item it would have been added when you checked the previous one :)